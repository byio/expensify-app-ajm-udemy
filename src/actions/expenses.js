import uuid from 'uuid';
import db from '../firebase/firebase';

export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    // return the promise chain below so that we can do promise chaining in the test suite for this method
    return db.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    })
  };
};

export const removeExpense = ({
  id
} = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// SET_EXPENSES (set the array value that we get back from firebase)
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// START_SET_EXPENSES (async method that calls setExpenses defined above)
export const startSetExpenses = () => {
  return (dispatch) => {
    return db.ref('expenses')
      .once('value')
      .then((snapshot) => {
        const expenses = [];
        snapshot.forEach((childSnapshot) => {
          expenses.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
