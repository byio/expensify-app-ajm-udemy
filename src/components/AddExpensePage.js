import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  handleExpenseForm = (expense) => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  };
  render () {
    return (
      <div>
        From Add Expense Page.
        <ExpenseForm
          handleExpenseForm={this.handleExpenseForm}
        />
      </div>
    );
  };
}

// const mapDispatchToProps = (dispatch) => {
//   return {
//     addExpense: (expense) => dispatch(addExpense(expense)) //the name of the returned method should ideally match the action generator
//   };
// };

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
