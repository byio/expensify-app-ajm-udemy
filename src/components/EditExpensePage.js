import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  handleExpenseForm = (expense) => {
    // this.props.dispatch(editExpense(this.props.expense.id, expense));
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  handleRemoveClick = () => {
    // this.props.dispatch(removeExpense({ id: this.props.expense.id }));
    this.props.removeExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  render () {
    return (
      <div>
        {/* From Edit Expense Page. Editing expense with Id = {this.props.match.params.id}. */}
        {console.log('this.props.match: ', this.props.match)}
        <ExpenseForm
          expense={this.props.expense}
          handleExpenseForm={this.handleExpenseForm}
        />
        <button onClick={this.handleRemoveClick}>
          Remove
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, expense) => dispatch(editExpense(id, expense)),
  removeExpense: (data) => dispatch(removeExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
