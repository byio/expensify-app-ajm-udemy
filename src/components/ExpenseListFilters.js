import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter,sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
// import 'react-dates/lib/css/_datepicker.css';

export class ExpenseListFilters extends React.Component {
  state = {
    calFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calFocused) => {
    this.setState(() => ({ calFocused }));
  };
  onSearchTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };
  render () {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input className="text-input" placeholder="Search expenses" type="text" value={this.props.filters.text} onChange={this.onSearchTextChange}/>
          </div>
          <div className="input-group__item">
            <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={() => false}
              showClearDates={true}
            />
          </div>
        </div>
      </div>
    );
  }
}

// const mapStateToProps = (state) => {
//   return {
//     filters: state.filters
//   };
// };

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
