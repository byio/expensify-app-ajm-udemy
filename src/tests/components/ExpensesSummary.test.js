import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

// test case: single expense
test('should render expenses summary correctly with 1 expense', () => {
  const wrapper = shallow(<ExpensesSummary
    expensesCount={1}
    expensesTotal={197}
  />);
  expect(wrapper).toMatchSnapshot();
});

// test case: multiple expenses
test('should render expenses summary correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary
    expensesCount={23}
    expensesTotal={156795}
  />);
  expect(wrapper).toMatchSnapshot();
});

// test case: zero expenses
test('should render expenses summary corectly with zero expneses', () => {
  const wrapper = shallow(<ExpensesSummary
    expensesCount={0}
    expensesTotal={0}
  />);
  expect(wrapper).toMatchSnapshot();
});
