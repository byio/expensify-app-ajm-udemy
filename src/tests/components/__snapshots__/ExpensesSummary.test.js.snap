// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expenses summary corectly with zero expneses 1`] = `
<div>
  <p>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
    .
  </p>
</div>
`;

exports[`should render expenses summary correctly with 1 expense 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $1.97
    .
  </p>
</div>
`;

exports[`should render expenses summary correctly with multiple expenses 1`] = `
<div>
  <p>
    Viewing 
    23
     
    expenses
     totalling 
    $1,567.95
    .
  </p>
</div>
`;
